test/js-api/gc/casts.tentative.any.js

```diff
@@ -0,0 +1,303 @@
+  assert_throws_js(TypeError, () => exports.arrayNullableArg(42));
+  assert_throws_js(TypeError, () => exports.arrayNullableArg(42n));
+  assert_throws_js(TypeError, () => exports.arrayNullableArg("foo"));
+  assert_throws_js(TypeError, () => exports.arrayNullableArg({}));
+  assert_throws_js(TypeError, () => exports.arrayNullableArg(exports.testFunc));
+  assert_throws_js(TypeError, () => exports.arrayNullableArg(() => {}));
+}, "arrayref casts");
+
+test(() => {
+  exports.i31Arg(42);
+  assert_throws_js(TypeError, () => exports.i31Arg(exports.makeStruct()));
+  assert_throws_js(TypeError, () => exports.i31Arg(exports.makeArray()));
+  assert_throws_js(TypeError, () => exports.i31Arg(42n));
+  assert_throws_js(TypeError, () => exports.i31Arg("foo"));
+  assert_throws_js(TypeError, () => exports.i31Arg({}));
+  assert_throws_js(TypeError, () => exports.i31Arg(exports.testFunc));
+  assert_throws_js(TypeError, () => exports.i31Arg(() => {}));
+  assert_throws_js(TypeError, () => exports.i31Arg(null));
+
+  exports.i31NullableArg(null);
+  exports.i31NullableArg(42);
+  assert_throws_js(TypeError, () => exports.i31NullableArg(exports.makeStruct()));
+  assert_throws_js(TypeError, () => exports.i31NullableArg(exports.makeArray()));
+  assert_throws_js(TypeError, () => exports.i31NullableArg(42n));
+  assert_throws_js(TypeError, () => exports.i31NullableArg("foo"));
+  assert_throws_js(TypeError, () => exports.i31NullableArg({}));
+  assert_throws_js(TypeError, () => exports.i31NullableArg(exports.testFunc));
+  assert_throws_js(TypeError, () => exports.i31NullableArg(() => {}));
+}, "i31ref casts");
+
+test(() => {
+  exports.funcArg(exports.testFunc);
+  assert_throws_js(TypeError, () => exports.funcArg(exports.makeStruct()));
+  assert_throws_js(TypeError, () => exports.funcArg(exports.makeArray()));
+  assert_throws_js(TypeError, () => exports.funcArg(42));
+  assert_throws_js(TypeError, () => exports.funcArg(42n));
+  assert_throws_js(TypeError, () => exports.funcArg("foo"));
+  assert_throws_js(TypeError, () => exports.funcArg({}));
+  assert_throws_js(TypeError, () => exports.funcArg(() => {}));
+  assert_throws_js(TypeError, () => exports.funcArg(null));
+
+  exports.funcNullableArg(null);
+  exports.funcNullableArg(exports.testFunc);
+  assert_throws_js(TypeError, () => exports.funcNullableArg(exports.makeStruct()));
+  assert_throws_js(TypeError, () => exports.funcNullableArg(exports.makeArray()));
+  assert_throws_js(TypeError, () => exports.funcNullableArg(42));
+  assert_throws_js(TypeError, () => exports.funcNullableArg(42n));
+  assert_throws_js(TypeError, () => exports.funcNullableArg("foo"));
+  assert_throws_js(TypeError, () => exports.funcNullableArg({}));
+  assert_throws_js(TypeError, () => exports.funcNullableArg(() => {}));
+}, "funcref casts");
+
+test(() => {
+  for (const nullfunc of [exports.noneArg, exports.nofuncArg, exports.noexternArg]) {
+    assert_throws_js(TypeError, () => nullfunc(exports.makeStruct()));
```

Comment:
I added a few more test cases as I realized tests for bottom types were missing.

Here, V8 (commit 014e0bb5ac4691c85b1da2dac579af77d02d17b7) does not throw while Spidermonkey and JSC do. This test is checking all of the bottom types, and in V8 the failure is just for `noextern`.

I believe casting to `noextern` should fail for anything but null (if it's a nullable ref).

Reply:
@jakobkummerow could the above be a V8 bug?

Reply:
> I believe casting to noextern should fail for any of the arguments but null (if it's a nullable ref, as in the loop below this one).

I think you're right, but the spec is wrong about that: on the extern ref path, it skips the subtype check, which is wrong. Please see #442 for a fix.